refactor!: split `EmbodiedEnvironment` into protocols (#672)
thousandbrainsproject/tbp.monty: docs/how-to-use-monty/tutorials/using-monty-in-a-custom-application.md

Reviewer: **suggestion**:
```suggestion
For example, if you have one agent with two sensors, the proprioceptive state could look like this:
```

Additionally, the `EnvironmentInterface` and Environment can implement specific functions to be executed at different points in the experiment, such as resetting the agent position and showing a new object or scene at the beginning of a new episode.

> To use Monty in a custom environment, you usually need to customize the `EnvironmentInterface` and implement the `SimulatedObjectEnvironment` protocol in custom classes. For example, if you look back at the previous tutorials, you will see that for those Habitat experiments, we've been using the `EnvironmentInterfacePerObject` and the `HabitatEnvironment`. The diagram below shows some key elements that need to be defined for these two classes. It's best to start thinking about the environment setup first, as this will force you to think through how to structure your application correctly for Monty to tackle.

Reviewer: **suggestion**:

```suggestion
To use Monty in a custom environment, you usually need to customize the `EnvironmentInterface` class and write a custom implementation of the `SimulatedObjectEnvironment` protocol. For example, if you look back at the previous tutorials, you will see that for those Habitat experiments, we've been using the `EnvironmentInterfacePerObject` and the `HabitatEnvironment`. The diagram below shows some key elements that need to be defined for these two classes. It's best to start thinking about the environment setup first, as this will force you to think through how to structure your application correctly for Monty to tackle.
```

If your application requires a specific policy to move through the environment or you have a complex actuator to control, you might want to implement a custom `MotorSystem` or `MotorPolicy` class. For more details on our existing motor system and policies, see our [documentation on Monty's policies](https://thousandbrainsproject.readme.io/docs/policy).

Writing those custom classes works the same way as it does for the `EnvironmentInterface` and `Environment` protocols. For general information, see our documentation on [customizing Monty](https://thousandbrainsproject.readme.io/docs/customizing-monty).

# Conclusion

> This tutorial was a bit more text than practical code. This is because every application is different, and we try to convey the general principles here. The first step for any application is to think about if and how the task can be phrased as a sensorimotor environment. What is Monty's action space? How is movement defined? How does it change observations? How do movement and sensation determine the sensor's location and orientation in space? This will then help you figure out how to define a custom `Environment` and `EnvironmentInterface`, and their associated `__init__`, `step`, `get_state`,`reset`, `pre_episode`, and `post_episode` methods. If you run into issues customizing Monty to your application, please come over to our [Discourse Forum](https://thousandbrains.discourse.group/) and ask for help!

Reviewer: **suggestion (non-blocking)**: `get_state` is no longer part of the framework, although it is an internal convenience method in some classes.
```suggestion
This tutorial was a bit more text than practical code. This is because every application is different, and we try to convey the general principles here. The first step for any application is to think about if and how the task can be phrased as a sensorimotor environment. What is Monty's action space? How is movement defined? How does it change observations? How do movement and sensation determine the sensor's location and orientation in space? This will then help you figure out how to define a custom `Environment` and `EnvironmentInterface`, and their associated `__init__`, `step`,`reset`, `pre_episode`, and `post_episode` methods. If you run into issues customizing Monty to your application, please come over to our [Discourse Forum](https://thousandbrains.discourse.group/) and ask for help!
```
